serve/mlc_serve/model/sampler.py

```diff
@@ -372,20 +382,39 @@ def from_sampling_params(
         )
 
 
-def adjust_logits(logits, sampling_metadata, vocab_size):
+def get_bin_counts_and_mask(
```

Comment:
Valery talked about this in [the issue](https://linear.app/octoai/issue/DEE-64/fix-repetition-penalty-to-correctly-account-for-token-history-and), repasting some of it here:
> In my point of view the good place is SamplingState, but it calculates each time for new request. It looks like it is not good thing due to in general it can be done once, but redo this requires much time. May be SamplingParams in Request should be replaced by SamplingState with some sharing for parameter n > 1.
> What do you think about it?
> Just now I plan to save it in SamplingParams for quick fix, but we need a better place

Reply:
Ah, thanks Jason and this is great point. @vvchernov, can we follow-up about this? We can separate what needs/does not need to be updated each iteration like you raised.

Reply:
Oh, thanks @binarybana to share my ideas! I repeat it again in other words above.

> can we follow-up about this? We can separate what needs/does not need to be updated each iteration like you raised.

First of all it looks like a separate task, if we need to support correct repetition penalty in high priority it is better to do it first. I see the task is redesign of API, it should be done carefully and we need more discussions in details. I can prepare separate draft of PR with some dirty code which implements my idea and we will discuss there how it should be. And I'm still aware about logits_processor. Now it looks like it was done for json mode, but in general view it does the same as sampler. It is better also rethink this moment in our redesign